### Optimizing A Data Set for Memory Usage

In [5]:
import pandas as pd

In [6]:
employees_data = '/Users/ypushiev/Learning/PANDAS IN ACTION/Chapter 5 Dataframe filtering/Data/employees.csv'

In [7]:
pd.read_csv(employees_data).isna().sum()

First Name     68
Gender        147
Start Date      2
Salary          2
Mgmt           68
Team           44
dtype: int64

In [44]:
df_employees = pd.read_csv(employees_data, parse_dates = ['Start Date'])

/var/folders/pg/x6kyk1l92fj8q4332g1j8rnw0000gp/T/ipykernel_81492/4257891297.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_employees = pd.read_csv(employees_data, parse_dates = ['Start Date'])


In [45]:
df_employees['Start Date'] = pd.to_datetime(df_employees['Start Date'], format='%Y-%m-%d')

In [46]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   First Name  933 non-null    object        
 1   Gender      854 non-null    object        
 2   Start Date  999 non-null    datetime64[ns]
 3   Salary      999 non-null    float64       
 4   Mgmt        933 non-null    object        
 5   Team        957 non-null    object        
dtypes: datetime64[ns](1), float64(1), object(4)
memory usage: 47.1+ KB


**memory usage: 47.1+ KB**

In [47]:
df_employees.head()

First Name  Gender Start Date    Salary   Mgmt       Team
0    Douglas    Male 1993-08-06       NaN   True  Marketing
1     Thomas    Male 1996-03-31   61933.0   True        NaN
2      Maria  Female        NaT  130590.0  False    Finance
3      Jerry     NaN 2005-03-04  138705.0   True    Finance
4      Larry    Male 1998-01-24  101004.0   True         IT

**Convert datetime into the object to cut the time from the data**

In [48]:
df_employees['Start Date'] = df_employees['Start Date'].dt.strftime('%Y-%m-%d')

In [49]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   First Name  933 non-null    object 
 1   Gender      854 non-null    object 
 2   Start Date  999 non-null    object 
 3   Salary      999 non-null    float64
 4   Mgmt        933 non-null    object 
 5   Team        957 non-null    object 
dtypes: float64(1), object(5)
memory usage: 47.1+ KB


**memory usage: 47.1+ KB**

### Converting Data Types with the astype Method

In [50]:
df_employees['Mgmt'].astype('bool').tail()

996     False
997     False
998     False
999      True
1000     True
Name: Mgmt, dtype: bool

**Convert object type into the bool type and see results in the size**

In [51]:
df_employees['Mgmt'] = df_employees['Mgmt'].astype('bool')

In [52]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   First Name  933 non-null    object 
 1   Gender      854 non-null    object 
 2   Start Date  999 non-null    object 
 3   Salary      999 non-null    float64
 4   Mgmt        1001 non-null   bool   
 5   Team        957 non-null    object 
dtypes: bool(1), float64(1), object(4)
memory usage: 40.2+ KB


**memory usage: 40.2+ KB**

### Convert into the integer type

**The Salary column contains NaN values**

In [53]:
df_employees['Salary'].isna().sum()

2

**Replace NaN values by 0**

In [54]:
df_employees['Salary']= df_employees['Salary'].fillna(0)

In [56]:
df_employees['Salary']=df_employees['Salary'].astype('int')

In [57]:
df_employees['Salary'].head()

0         0
1     61933
2    130590
3    138705
4    101004
Name: Salary, dtype: int64

In [37]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   First Name  933 non-null    object
 1   Gender      854 non-null    object
 2   Start Date  999 non-null    object
 3   Salary      1001 non-null   int64 
 4   Mgmt        1001 non-null   bool  
 5   Team        957 non-null    object
dtypes: bool(1), int64(1), object(4)
memory usage: 40.2+ KB


**The memory might be reduced more using INT32**

**Check max value**

In [65]:
df_employees['Salary'].max()

149908

In [59]:
df_employees['Salary']=df_employees['Salary'].astype('int32')

In [60]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   First Name  933 non-null    object
 1   Gender      854 non-null    object
 2   Start Date  999 non-null    object
 3   Salary      1001 non-null   int32 
 4   Mgmt        1001 non-null   bool  
 5   Team        957 non-null    object
dtypes: bool(1), int32(1), object(4)
memory usage: 36.3+ KB


In [66]:
df_employees['Salary'].max()

149908

In [61]:
df_employees['Salary'].head()

0         0
1     61933
2    130590
3    138705
4    101004
Name: Salary, dtype: int32

**memory usage: 36.3+ KB**

### Convert into categorical type

**Check unique values in the Dataframe** 

In [67]:
df_employees.nunique()

First Name    200
Gender          2
Start Date    971
Salary        995
Mgmt            2
Team           10
dtype: int64

**The Gender column contains only 2 unique values** 

In [69]:
df_employees['Gender'].astype('category').head()

0      Male
1      Male
2    Female
3       NaN
4      Male
Name: Gender, dtype: category
Categories (2, object): ['Female', 'Male']

In [70]:
df_employees['Gender'] = df_employees['Gender'].astype('category')

**Convert Team column type**

In [73]:
df_employees['Team'] = df_employees['Team'].astype('category')

In [74]:
df_employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1001 entries, 0 to 1000
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   First Name  933 non-null    object  
 1   Gender      854 non-null    category
 2   Start Date  999 non-null    object  
 3   Salary      1001 non-null   int32   
 4   Mgmt        1001 non-null   bool    
 5   Team        957 non-null    category
dtypes: bool(1), category(2), int32(1), object(2)
memory usage: 23.1+ KB


**memory usage: 23.1+ KB**

**The Dataframe size has been decreased from 47.1+ KB to 23.1+ KB approximately 50%**

### Filtering by a Single Condition

In [83]:
"Maria" == "Maria"

True

In [84]:
"Maria" == "Taylor"

False

**Check the column First Name**

In [85]:
df_employees["First Name"] == "Maria"

0       False
1       False
2        True
3       False
4       False
        ...  
996     False
997     False
998     False
999     False
1000    False
Name: First Name, Length: 1001, dtype: bool

**Select rows where "First Name" == "Maria"**

In [86]:
df_employees[df_employees["First Name"] == "Maria"]

First Name  Gender  Start Date  Salary   Mgmt          Team
2        Maria  Female         NaN  130590  False       Finance
198      Maria  Female  1990-12-27   36067   True       Product
815      Maria     NaN  1986-01-18  106562  False            HR
844      Maria     NaN  1985-06-19  148857  False         Legal
936      Maria  Female  2003-03-14   96250  False  Business Dev
984      Maria  Female  2011-10-15   43455  False   Engineering

**Create a filter for Dataframe**

In [87]:
marias = df_employees["First Name"] == "Maria"
df_employees[marias]

First Name  Gender  Start Date  Salary   Mgmt          Team
2        Maria  Female         NaN  130590  False       Finance
198      Maria  Female  1990-12-27   36067   True       Product
815      Maria     NaN  1986-01-18  106562  False            HR
844      Maria     NaN  1985-06-19  148857  False         Legal
936      Maria  Female  2003-03-14   96250  False  Business Dev
984      Maria  Female  2011-10-15   43455  False   Engineering

**Create a filter and find Top 5 Salary > 100000**

In [88]:
high_earners = df_employees["Salary"] > 100000
df_employees[high_earners].sort_values(by="Salary", ascending=False).head()

First Name  Gender  Start Date  Salary   Mgmt     Team
644  Katherine  Female  1996-08-13  149908  False  Finance
429       Rose  Female  2015-05-28  149903  False       HR
828    Cynthia  Female  2006-07-12  149684  False  Product
186        NaN  Female  2005-02-23  149654   True    Sales
160      Kathy  Female  2000-03-18  149563   True  Finance

### Filtering by Multiple Conditions

#### Condition and - &

**First condition**

In [89]:
is_female = df_employees['Gender'] == 'Female'

**Second condiotion**

In [92]:
in_biz_dev = df_employees['Team'] == "Business Dev"

**Find only women from the Business Development team**

In [94]:
df_employees[is_female & in_biz_dev].head()

First Name  Gender  Start Date  Salary   Mgmt          Team
9     Frances  Female  2002-08-08  139852   True  Business Dev
33       Jean  Female  1993-12-18  119082  False  Business Dev
36     Rachel  Female  2009-02-16  142032  False  Business Dev
38  Stephanie  Female  1986-09-13   36844   True  Business Dev
61     Denise  Female  2001-11-06  106862  False  Business Dev

In [95]:
is_manager = df_employees["Mgmt"]

**Add more conditionals**

In [96]:
df_employees[is_female & in_biz_dev & is_manager].head()

First Name  Gender  Start Date  Salary  Mgmt          Team
9      Frances  Female  2002-08-08  139852  True  Business Dev
38   Stephanie  Female  1986-09-13   36844  True  Business Dev
66       Nancy  Female  2012-12-15  125250  True  Business Dev
92       Linda  Female  2000-05-25  119009  True  Business Dev
111     Bonnie  Female  1999-12-17   42153  True  Business Dev

#### Condiotional OR - |